File: windows/Vinculum.Rendering/VinculumWire.cs
using System.Buffers.Binary;

namespace Vinculum.Rendering;

/// Decodes the VDL1 wire format (docs/DISPLAYLIST.md) — the C# twin of Swift
/// DisplayListWire and Kotlin VinculumWire. Consumes an UNTRUSTED buffer: returns
/// null (never throws) on bad magic, truncation, unknown tag, or a forged count;
/// never allocates on an untrusted length.
public static class VinculumWire
{
    public static DisplayList? Decode(byte[] b)
    {
        if (b.Length < 20) return null;
        if (b[0] != (byte)'V' || b[1] != (byte)'D' || b[2] != (byte)'L' || b[3] != (byte)'1') return null;
        try
        {
            int p = 4;
            float F() { var v = BinaryPrimitives.ReadSingleLittleEndian(b.AsSpan(p)); p += 4; return v; }
            uint U() { var v = BinaryPrimitives.ReadUInt32LittleEndian(b.AsSpan(p)); p += 4; return v; }
            byte B() => b[p++];
            uint RGBA() { uint r = B(), g = B(), bl = B(), a = B(); return (a << 24) | (r << 16) | (g << 8) | bl; }

            List<Seg> Segs()
            {
                uint n = U();
                if (n > (uint)(b.Length - p + 1)) throw new InvalidOperationException("bad segCount");
                var segs = new List<Seg>((int)Math.Min(n, (uint)b.Length));
                for (uint i = 0; i < n; i++)
                {
                    byte so = B();
                    segs.Add(so switch
                    {
                        0 or 1 => new Seg((SegOp)so, new[] { F(), F() }),
                        2 => new Seg(SegOp.Quad, new[] { F(), F(), F(), F() }),
                        3 => new Seg(SegOp.Cubic, new[] { F(), F(), F(), F(), F(), F() }),
                        4 => new Seg(SegOp.Close, Array.Empty<float>()),
                        _ => throw new InvalidOperationException("bad seg op"),
                    });
                }
                return segs;
            }

            float width = F(), ascent = F(), descent = F();
            uint opCount = U();
            if (opCount > (uint)b.Length) return null;   // untrusted count
            var ops = new List<Op>((int)Math.Min(opCount, (uint)b.Length));
            for (uint i = 0; i < opCount; i++)
            {
                switch (B())
                {
                    case 0: { uint c = RGBA(); ops.Add(new FillPath(Segs(), c)); break; }   // rgba THEN segs
                    case 1: { uint c = RGBA(); ops.Add(new FillRect(F(), F(), F(), F(), c)); break; }
                    case 2: { uint c = RGBA(); float w = F(); byte cap = B(), join = B(); ops.Add(new StrokePath(Segs(), w, cap, join, c)); break; }
                    default: return null;
                }
            }
            return new DisplayList(width, ascent, descent, ops);
        }
        catch { return null; }
    }
}

File: android/vinculum/src/main/kotlin/ai/vinc/VinculumWire.kt
package ai.vinc

import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Decodes the `VDL1` wire format (Swift `DisplayListWire`) into a drawable
 * display list. The Kotlin counterpart of the Swift serializer — tested against
 * the same committed fixture, so the two ends cannot drift.
 *
 * Coordinates are scene-native: y-up, origin on the baseline. The renderer
 * applies the y-flip once, at draw time.
 */
object VinculumWire {

    /** A concrete drawing primitive: filled outline, filled rect, or stroke. */
    sealed interface Op {
        data class FillPath(val subpaths: List<Seg>, val color: Int) : Op
        data class FillRect(val x: Float, val y: Float, val w: Float, val h: Float, val color: Int) : Op
        data class StrokePath(
            val path: List<Seg>, val width: Float, val cap: Int, val join: Int, val color: Int
        ) : Op
    }

    /** A path segment. `op`: 0 move, 1 line, 2 quad, 3 cubic, 4 close. */
    class Seg(val op: Int, val pts: FloatArray)

    class DisplayList(
        val width: Float, val ascent: Float, val descent: Float, val ops: List<Op>
    ) { val height: Float get() = ascent + descent }

    /** Decodes VDL1 bytes, or null on a bad magic / truncation (never throws). */
    fun decode(bytes: ByteArray): DisplayList? {
        if (bytes.size < 20) return null
        val b = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN)
        // magic "VDL1"
        if (b.get() != 'V'.code.toByte() || b.get() != 'D'.code.toByte() ||
            b.get() != 'L'.code.toByte() || b.get() != '1'.code.toByte()) return null
        return try {
            val width = b.float; val ascent = b.float; val descent = b.float
            val opCount = b.int
            if (opCount < 0 || opCount > bytes.size) return null   // untrusted count
            val ops = ArrayList<Op>(minOf(opCount, bytes.size))
            repeat(opCount) {
                when (b.get().toInt()) {
                    0 -> { val c = rgba(b); ops.add(Op.FillPath(readSegs(b), c)) }  // rgba THEN segs
                    1 -> { val c = rgba(b); ops.add(Op.FillRect(b.float, b.float, b.float, b.float, c)) }
                    2 -> {
                        val c = rgba(b); val w = b.float
                        val cap = b.get().toInt(); val join = b.get().toInt()
                        ops.add(Op.StrokePath(readSegs(b), w, cap, join, c))
                    }
                    else -> return null
                }
            }
            DisplayList(width, ascent, descent, ops)
        } catch (_: Exception) { null }
    }

    private fun readSegs(b: ByteBuffer): List<Seg> {
        val n = b.int
        if (n < 0 || n > b.remaining() + 1) throw IllegalStateException("bad segCount")
        val segs = ArrayList<Seg>(minOf(n, b.remaining() + 1))
        repeat(n) {
            when (val op = b.get().toInt()) {
                0, 1 -> segs.add(Seg(op, floatArrayOf(b.float, b.float)))
                2 -> segs.add(Seg(op, floatArrayOf(b.float, b.float, b.float, b.float)))
                3 -> segs.add(Seg(op, floatArrayOf(b.float, b.float, b.float, b.float, b.float, b.float)))
                4 -> segs.add(Seg(op, FloatArray(0)))
                else -> throw IllegalStateException("bad seg op")
            }
        }
        return segs
    }

    /** rgba bytes → an ARGB int for android.graphics.Color. */
    private fun rgba(b: ByteBuffer): Int {
        val r = b.get().toInt() and 0xFF
        val g = b.get().toInt() and 0xFF
        val bl = b.get().toInt() and 0xFF
        val a = b.get().toInt() and 0xFF
        return (a shl 24) or (r shl 16) or (g shl 8) or bl
    }
}
